### PR TITLE
[node-manager] Update autoscaler

### LINKED ISSUE #2020: [node-manager] Advanced scheduling for cluster-autoscaler
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Use case. Why is this important?

1. There can be cases when after several cycles of node groups scale down and scale up, we have such state
```
["application-nodes","TEST_AZ_3",1]
["application-nodes","TEST_AZ_2",4]
["application-nodes","TEST_AZ_1",1]
```
This looks like a potential problem, if the zone with the largest number of nodes goes down.

2. There is no state sync between cluster-autoscaler and machine-controller-manager. For example, cluster-autoscaler decided to scale node group `my-node` with type `super`. If machine-controller-manager can't order machine with type `super`, cluster-autoscaler doesn't try to scale up other group. This happens because cluster-autoscale knows nothing about machine order status.

### Proposed Solution

- [x] Check if `--balance-similar-node-groups` option in cluster-autoscaler solves the problem (#2735)
- [ ] Fix scaling from zero and add hook in Deckhouse that helps cluster-autoscaler to reconsider which node group to scale (#2762)

### Additional Information

_No response_

### LINKED ISSUE #2473: NodeGroups priorities
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Use case. Why is this important?

There are some use cases that are not covered by the current node groups logic:
- spots are over in one zone, so we need to order them in another zone;
- the cheapest spots are over, so we need to order the more expensive ones.

### Proposed Solution

It would be nice to have some sort of priorities among the node groups.

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by yalosev**
## Description
Enable `--balance-similar-node-groups` to automatically balance nodes on scale-up between zones. This mechanism is described in details [here](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#im-running-cluster-with-nodes-in-multiple-zones-for-ha-purposes-is-that-supported-by-cluster-autoscaler). In general it will keep more balanced zones

Realize priority for cloud node zones. At first, CA tries to order node from a node group with the highest priority (if set), if node is not available to be ordered in 15 minutes - fallback to the next priority

Close #2020 
Close #2473 
Ref #2078 

## Why do we need it, and what problem does it solve?
Keep more balanced zones and get flexible mechanism for ordering nodes.

For example:
Adding 2 more pods (losev-worker-99e6104c node group already has 2 nodes):
```
I1016 18:47:35.130165       1 klogx.go:86] Pod default/redis-cache-56cd95799f-hqfv7 is unschedulable
I1016 18:47:35.130171       1 klogx.go:86] Pod default/redis-cache-56cd95799f-q48qh is unschedulable
I1016 18:47:35.131825       1 priority.go:163] priority expander: losev-worker-99e6104c chosen as the highest available
I1016 18:47:35.131839       1 priority.go:163] priority expander: losev-worker-e7ca5e66 chosen as the highest available
I1016 18:47:35.131845       1 waste.go:55] Expanding Node Group losev-worker-99e6104c would waste 100.00% CPU, 100.00% Memory, 100.00% Blended
I1016 18:47:35.131854       1 waste.go:55] Expanding Node Group losev-worker-e7ca5e66 would waste 100.00% CPU, 100.00% Memory, 100.00% Blended
I1016 18:47:35.131867       1 scale_up.go:468] Best option to resize: losev-worker-e7ca5e66
I1016 18:47:35.131874       1 scale_up.go:472] Estimated 2 nodes needed in losev-worker-e7ca5e66
I1016 18:47:35.131919       1 scale_up.go:585] Splitting scale-up between 2 similar node groups: {losev-worker-e7ca5e66, losev-worker-99e6104c}
I1016 18:47:35.131946       1 scale_up.go:595] Final scale-up plan: [{losev-worker-e7ca5e66 1->3 (max: 3)}]
I1016 18:47:35.131958       1 scale_up.go:691] Scale-up: setting group losev-worker-e7ca5e66 size to 3
```

---
NodeGroup with the highest priority is not able to join a node in a 15 minutes. Fallback to the next one.
`worker-big` has priority `50`
`worker` has priority `30`
```
I1017 16:32:58.842640       1 static_autoscaler.go:603] Decreasing size of losev-worker-big-54b5a293, expected=2 current=1 delta=-1

I1017 16:33:08.865182       1 klogx.go:86] Pod default/redis-cache-669897455-8rldz is unschedulable
W1017 16:33:08.868375       1 scale_up.go:399] Node group losev-worker-big-54b5a293 is not ready for scaleup - backoff
I1017 16:33:08.868560       1 priority.go:163] priority expander: losev-worker-54b5a293 chosen as the highest available
I1017 16:33:08.868579       1 scale_up.go:468] Best option to resize: losev-worker-54b5a293
I1017 16:33:08.868590       1 scale_up.go:472] Estimated 1 nodes needed in losev-worker-54b5a293
I1017 16:33:08.868658       1 scale_up.go:595] Final scale-up plan: [{losev-worker-54b5a293 2->3 (max: 4)}]
I1017 16:33:08.868680       1 scale_up.go:691] Scale-up: setting group losev-worker-54b5a293 size to 3
```

## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: node-manager
type: feature
summary: Cluster-autoscaler: smart balancer between zones in a node-group. CA will try to equal nodes count between zones during the scale-up. New priority field in the node-group spec, which can set order of scaling up nodes between different node-groups
```

### CODE REVIEW

**Review by z9r5 — APPROVED** [review]

File: modules/040-node-manager/hooks/internal/v1alpha2/nodegroup.go [hunk 1]
```diff
@@ -133,6 +133,9 @@ type CloudInstances struct {
 
 	// Reference to a ClassInstance resource. Required.
 	ClassReference ClassReference `json:"classReference"`
+
+	// Priority setting for autoscaler expander
+	Priority *int32 `json:"priority,omitempty"`
```
Comment (RomanenkoDenys) [inline modules/040-node-manager/hooks/internal/v1alpha2/nodegroup.go:138]: Why we need this field in v1alpha2 ?

File: modules/040-node-manager/hooks/set_ng_priorities_test.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+		stateNGs = `
+---
+apiVersion: deckhouse.io/v1
+kind: NodeGroup
+metadata:
+  name: ng1
+spec:
+  cloudInstances:
+    maxPerZone: 4
+    minPerZone: 3 # "$replicas" == "null"
```
Comment (RomanenkoDenys) [inline modules/040-node-manager/hooks/set_ng_priorities_test.go:37]: remove $replicas comment please

**Review by RomanenkoDenys — COMMENTED** [review]
LGTM

File: modules/040-node-manager/hooks/internal/v1alpha2/nodegroup.go (hunk 1 above)
Comment (yalosev) [reply modules/040-node-manager/hooks/internal/v1alpha2/nodegroup.go]: Removed

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Review by RomanenkoDenys — APPROVED** [review]
LGTM

**Review by nabokihms — APPROVED** [review]

**Pull request merged**